<a href="https://colab.research.google.com/github/SNEHA-17/ETE456_LabProjects/blob/main/1608004_Regression_Project_01.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

#**ETE456: Implementation of Regression Algorithms on a Real World Dataset**

**Objectives:**


1.   To pre-process the data before feeding into the ML models.
2.   To apply various regression algorithms on the real world dataset.



**Problem Statement:**
Accurately predicting the insurance costs from the given dataset.

**Data Preprocessing:**



*   Taking care of missing data
*   label encoding and then implementing on different models to find R-squared values. 


*   One Hot encoding and then implementing on different models with calculation of R-squared values.








In [ ]:
# Downloading the data
!wget -O insurance.csv https://www.dropbox.com/s/mwgqgjbmfw0xa5p/insurance.csv?dl=0

--2021-12-01 11:13:27--  https://www.dropbox.com/s/mwgqgjbmfw0xa5p/insurance.csv?dl=0
Resolving www.dropbox.com (www.dropbox.com)... 162.125.5.18, 2620:100:601d:18::a27d:512
Connecting to www.dropbox.com (www.dropbox.com)|162.125.5.18|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: /s/raw/mwgqgjbmfw0xa5p/insurance.csv [following]
--2021-12-01 11:13:28--  https://www.dropbox.com/s/raw/mwgqgjbmfw0xa5p/insurance.csv
Reusing existing connection to www.dropbox.com:443.
HTTP request sent, awaiting response... 302 Found
Location: https://ucc424ccc1bc10f0eadb8142f54d.dl.dropboxusercontent.com/cd/0/inline/BbBu_PvECGF1vVvQIhJFvtSavzyVXSVomWXngkZCVOT4t45adMMRpibkSDKsgSZ4kHY2gHFXaIruhoRusbrj0gUny8Ll99ZHZ1Yd4tk5tTwVerIUmyK88X9oXha4LxEPOn37nAkYtlH-BeuUSQozdjxA/file# [following]
--2021-12-01 11:13:28--  https://ucc424ccc1bc10f0eadb8142f54d.dl.dropboxusercontent.com/cd/0/inline/BbBu_PvECGF1vVvQIhJFvtSavzyVXSVomWXngkZCVOT4t45adMMRpibkSDKsgSZ4kHY2gHFXaIruhoRusb

In [ ]:
import warnings
warnings.filterwarnings("ignore")

In [ ]:
import numpy as np        
import pandas as pd     

In [ ]:
"""importing the dataset """

d_set = pd.read_csv('insurance.csv')
d_set

age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1338 rows x 7 columns]

In [ ]:
## Extracting the features and target column

features = d_set[['age','bmi','children','smoker']] 
target = d_set[['charges']]   

In [ ]:
features

age     bmi  children smoker
0      19  27.900         0    yes
1      18  33.770         1     no
2      28  33.000         3     no
3      33  22.705         0     no
4      32  28.880         0     no
...   ...     ...       ...    ...
1333   50  30.970         3     no
1334   18  31.920         0     no
1335   18  36.850         0     no
1336   21  25.800         0     no
1337   61  29.070         0    yes

[1338 rows x 4 columns]

In [ ]:
target

charges
0     16884.92400
1      1725.55230
2      4449.46200
3     21984.47061
4      3866.85520
...           ...
1333  10600.54830
1334   2205.98080
1335   1629.83350
1336   2007.94500
1337  29141.36030

[1338 rows x 1 columns]

In [ ]:
#Taking care of missing data

from sklearn.impute import SimpleImputer

In [ ]:
imputer = SimpleImputer(missing_values=np.nan,strategy = "mean")  
imputer = imputer.fit(features[['age','bmi','children']]) 
features[['age','bmi','children']]= imputer.transform(features[['age','bmi','children']])
features

age     bmi  children smoker
0     19.0  27.900       0.0    yes
1     18.0  33.770       1.0     no
2     28.0  33.000       3.0     no
3     33.0  22.705       0.0     no
4     32.0  28.880       0.0     no
...    ...     ...       ...    ...
1333  50.0  30.970       3.0     no
1334  18.0  31.920       0.0     no
1335  18.0  36.850       0.0     no
1336  21.0  25.800       0.0     no
1337  61.0  29.070       0.0    yes

[1338 rows x 4 columns]

**Implementing Label Encoding**

In [ ]:
from sklearn.preprocessing import LabelEncoder,OneHotEncoder

In [ ]:
labelencoder_f = LabelEncoder()

features['smoker'] = labelencoder_f.fit_transform(features['smoker'])
features

age     bmi  children  smoker
0     19.0  27.900       0.0       1
1     18.0  33.770       1.0       0
2     28.0  33.000       3.0       0
3     33.0  22.705       0.0       0
4     32.0  28.880       0.0       0
...    ...     ...       ...     ...
1333  50.0  30.970       3.0       0
1334  18.0  31.920       0.0       0
1335  18.0  36.850       0.0       0
1336  21.0  25.800       0.0       0
1337  61.0  29.070       0.0       1

[1338 rows x 4 columns]

In [ ]:
features['new_smoker'] = d_set['smoker']
features

age     bmi  children  smoker new_smoker
0     19.0  27.900       0.0       1        yes
1     18.0  33.770       1.0       0         no
2     28.0  33.000       3.0       0         no
3     33.0  22.705       0.0       0         no
4     32.0  28.880       0.0       0         no
...    ...     ...       ...     ...        ...
1333  50.0  30.970       3.0       0         no
1334  18.0  31.920       0.0       0         no
1335  18.0  36.850       0.0       0         no
1336  21.0  25.800       0.0       0         no
1337  61.0  29.070       0.0       1        yes

[1338 rows x 5 columns]

In [ ]:
d_set

age     sex     bmi  children smoker     region      charges
0      19  female  27.900         0    yes  southwest  16884.92400
1      18    male  33.770         1     no  southeast   1725.55230
2      28    male  33.000         3     no  southeast   4449.46200
3      33    male  22.705         0     no  northwest  21984.47061
4      32    male  28.880         0     no  northwest   3866.85520
...   ...     ...     ...       ...    ...        ...          ...
1333   50    male  30.970         3     no  northwest  10600.54830
1334   18  female  31.920         0     no  northeast   2205.98080
1335   18  female  36.850         0     no  southeast   1629.83350
1336   21  female  25.800         0     no  southwest   2007.94500
1337   61  female  29.070         0    yes  northwest  29141.36030

[1338 rows x 7 columns]

In [ ]:
encoder=OneHotEncoder(sparse=False)
encoded_labels = pd.DataFrame (encoder.fit_transform(features[['smoker']]))
encoded_labels.columns = encoder.get_feature_names(['smoker'])
d_set= pd.concat([features, encoded_labels ], axis=1)
d_set

age     bmi  children  smoker new_smoker  smoker_0  smoker_1
0     19.0  27.900       0.0       1        yes       0.0       1.0
1     18.0  33.770       1.0       0         no       1.0       0.0
2     28.0  33.000       3.0       0         no       1.0       0.0
3     33.0  22.705       0.0       0         no       1.0       0.0
4     32.0  28.880       0.0       0         no       1.0       0.0
...    ...     ...       ...     ...        ...       ...       ...
1333  50.0  30.970       3.0       0         no       1.0       0.0
1334  18.0  31.920       0.0       0         no       1.0       0.0
1335  18.0  36.850       0.0       0         no       1.0       0.0
1336  21.0  25.800       0.0       0         no       1.0       0.0
1337  61.0  29.070       0.0       1        yes       0.0       1.0

[1338 rows x 7 columns]

In [ ]:
d_set.columns

Index(['age', 'bmi', 'children', 'smoker', 'new_smoker', 'smoker_0',
       'smoker_1'],
      dtype='object')

In [ ]:
## For label encoding features
n_features = d_set[['age','bmi', 'children','smoker']]

In [ ]:
## For one hot encoding features
n_features1 = d_set[['age','bmi', 'children','smoker_0',
       'smoker_1']]

**Splitting the dataset into train set and test set:**

In [ ]:
from sklearn.model_selection import train_test_split
X_train,X_test,y_train,y_test=train_test_split(n_features,target,test_size = 0.2,random_state = 0)

In [ ]:
print(X_train.shape)
print(X_test.shape)

(1070, 4)
(268, 4)


In [ ]:
X_train

age     bmi  children  smoker
621   37.0  34.100       4.0       1
194   18.0  34.430       0.0       0
240   23.0  36.670       2.0       1
1168  32.0  35.200       2.0       0
1192  58.0  32.395       1.0       0
...    ...     ...       ...     ...
763   27.0  26.030       0.0       0
835   42.0  35.970       2.0       0
1216  40.0  25.080       0.0       0
559   19.0  35.530       0.0       0
684   33.0  18.500       1.0       0

[1070 rows x 4 columns]

**Multivariate linear Regression Model with Label Encoding\:**

In [ ]:
# Fitting Multiple Linear Regression Model to Training Set
from sklearn.linear_model import LinearRegression

regressor = LinearRegression()
regressor.fit(X_train,y_train)

# Predicting the Test Set Results
y_pred = regressor.predict(X_test)

In [ ]:
y_test

charges
578    9724.53000
610    8547.69130
569   45702.02235
1034  12950.07120
198    9644.25250
...           ...
1084  15019.76005
726    6664.68595
1132  20709.02034
725   40932.42950
963    9500.57305

[268 rows x 1 columns]

**Finding Evaluation Metrix:**

**R-Squared Error:**



In [ ]:
from sklearn.metrics import r2_score

r2_score(y_test, y_pred)

0.7978274606303823

**Mean Squared Error:**

In [ ]:
from sklearn.metrics import mean_squared_error

mean_squared_error(y_test, y_pred)

32171708.57245572

**Support Vector Regression with Label Encoding:**

In [ ]:
# Fitting SVR to the dataset
from sklearn.svm import SVR 

regressor = SVR(kernel = 'linear')
regressor.fit(X_train,y_train)

SVR(kernel='linear')

In [ ]:
y_pred = regressor.predict(X_test)

In [ ]:
## R-squared error
r2_score(y_test, y_pred)

-0.08697011014493672

In [ ]:
## mean squared error
from sklearn.metrics import mean_squared_error

mean_squared_error(y_test, y_pred)

172969512.6726405

**Decision Tree Regression with Label Encoding:**

In [ ]:
# Fitting the decision tree regression with the dataset
from sklearn.tree import DecisionTreeRegressor


regressor = DecisionTreeRegressor(random_state = 0)
regressor.fit(X_train,y_train) 


DecisionTreeRegressor(random_state=0)

In [ ]:
## r squared error
y_pred = regressor.predict(X_test)
r2_score(y_test, y_pred)

0.7289056282757718

In [ ]:
## mean squared error
from sklearn.metrics import mean_squared_error

mean_squared_error(y_test, y_pred)

43139237.15821675

**Random Forest Regression with Label Encoding:**

In [ ]:
# Fitting the Random Forest Regression with the dataset
from sklearn.ensemble import RandomForestRegressor

regressor = RandomForestRegressor(n_estimators = 10,random_state = 0) # n estiamator is the number of decision trees
regressor.fit(X_train,y_train) 

RandomForestRegressor(n_estimators=10, random_state=0)

In [ ]:
## r squared error
y_pred = regressor.predict(X_test)
r2_score(y_test, y_pred)

0.8577220757841139

In [ ]:
## mean squared error
mean_squared_error(y_test, y_pred)

22640680.719744213

**Multivariate linear Regression Model with One Hot Encoding:**

In [ ]:
from sklearn.model_selection import train_test_split
X_train1,X_test1,y_train1,y_test1=train_test_split(n_features1,target,test_size = 0.2,random_state = 0)

# Fitting Multiple Linear Regression Model to Training Set
from sklearn.linear_model import LinearRegression

regressor = LinearRegression()
regressor.fit(X_train1,y_train1)

# Predicting the Test Set Results
y_pred1 = regressor.predict(X_test1)
y_test1

charges
578    9724.53000
610    8547.69130
569   45702.02235
1034  12950.07120
198    9644.25250
...           ...
1084  15019.76005
726    6664.68595
1132  20709.02034
725   40932.42950
963    9500.57305

[268 rows x 1 columns]

In [ ]:
from sklearn.metrics import r2_score

r2_score(y_test1, y_pred1)

0.7978274606303823

**Support Vector Regression with One Hot Encoding:**

In [ ]:
# Fitting SVR to the dataset
from sklearn.svm import SVR 

regressor = SVR(kernel = 'linear')
regressor.fit(X_train1,y_train1)
y_pred1 = regressor.predict(X_test1)

In [ ]:
## R-squared error
r2_score(y_test1, y_pred1)

-0.07245023797631767

**Decision Tree Regression with One Hot Encoding:**

In [ ]:
# Fitting the decision tree regression with the dataset
from sklearn.tree import DecisionTreeRegressor


regressor = DecisionTreeRegressor(random_state = 0)
regressor.fit(X_train1,y_train1) 

DecisionTreeRegressor(random_state=0)

In [ ]:
## r squared error
y_pred1 = regressor.predict(X_test1)
r2_score(y_test1, y_pred1)

0.7408884055687519

**Random Forest Regression with One Hot Encoding:**

In [ ]:
# Fitting the Random Forest Regression with the dataset
from sklearn.ensemble import RandomForestRegressor

regressor = RandomForestRegressor(n_estimators = 10,random_state = 0) 
regressor.fit(X_train1,y_train1) 

RandomForestRegressor(n_estimators=10, random_state=0)

In [ ]:
## r squared error
y_pred1 = regressor.predict(X_test1)
r2_score(y_test1, y_pred1)

0.8530672523889884

**Findings:**


*   With label Encoding R-squared values for different models: 

 

         1.   Multivariate Linear regression Model = approx 80%
         2.   Support Vector Regression = -8%
         3.   Decision Tree Regression = approx 73%
         4.   Random Forest regression = 86%







*   With label Encoding Mean-squared values for different models:
         1.   Multivariate Linear regression Model = 32171708.57
         2.   Support Vector Regression = 172969512.67
         3.   Decision Tree Regression = 43139237.16
         4.   Random Forest regression = 22640680.72
*  With One Hot Encoding R-squared values for different models: 
         1.   Multivariate Linear regression Model = 79.7%
         2.   Support Vector Regression = -7%
         3.   Decision Tree Regression = 74.1%
         4.   Random Forest regression = 85.31%






**Conclusion:**




From the above evaluation metrices results it is seen that, the Random Forest 
Regression model is best fitted for the given dataset both for label encoding and one hot encoding as R-Squared error is approximately 85-86% which is greater than from other models.